### ISSUE TITLE
W1b: `RandomAffine` and `RandomAffine3D` rotate opposite to every other kornia rotation entry point

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4408]
A positive angle means counter-clockwise as displayed almost everywhere in kornia: `rotate`, `Rotate`, `Affine`, `get_rotation_matrix2d` + `warp_affine`, `RandomRotation`, `RandomRotation90` and the `auto` `Rotate` operation all agree. `RandomAffine(degrees=+a)` and `get_affine_matrix2d(angle=+a)` turn the image the other way. The two that disagree are exactly the affine composer and the augmentation class built on it; the repro below reproduces six of these entry points directly, and `Rotate`, `Affine` and the `auto` `Rotate` behave as `rotate` does.

The negation is deliberate at the primitive: `get_affine_matrix2d` calls `get_rotation_matrix2d(center, -angle, scale)` and its `Convention:` block says so (`kornia/geometry/transform/imgwarp.py:940`). What is not stated anywhere is that `RandomAffine`, a user-facing augmentation, inherits it — so `RandomRotation(30)` and `RandomAffine(degrees=(30, 30))` on the same image produce different pictures, and a pipeline that swaps one for the other silently changes the sign of every rotation. Keypoints follow the image in both classes, so the two are each internally consistent; there is no single wrong pixel to point at, only two contradictory conventions.

The same split exists in 3D, mirrored on all three axes. `RandomRotation3D` reproduces `rotate3d` exactly for each of the three `degrees` slots; `RandomAffine3D` negates the in-plane offset of all three, because `get_affine_matrix3d` negates `angles` relative to `get_projective_transform` (`kornia/geometry/transform/imgwarp.py:1070`). The scope is angle-only: the same composer does **not** negate shear or translate, which is why `RandomShear` and `RandomAffine(shear=...)` build a byte-identical matrix and `RandomAffine`'s `M[0][2]`/`M[1][2]` equal the drawn translation unchanged.

## Repro

```python
import torch, kornia, kornia.augmentation as K
from kornia.geometry.transform import rotate, get_rotation_matrix2d, get_affine_matrix2d, warp_affine, rotate3d
print("kornia", kornia.__version__, "torch", torch.__version__)

def hot(h=7, w=7, r=1, c=4):
    x = torch.zeros(1, 1, h, w); x[0, 0, r, c] = 1.0; return x

x = hot()  # 7x7, hot pixel at (row 1, col 4); centre is (3, 3)
def argmax2d(t):
    i = int(t.flatten().argmax()); return (i // t.shape[-1], i % t.shape[-1])

torch.manual_seed(0)
print("RandomRotation(90)      argmax:", argmax2d(K.RandomRotation((90., 90.), p=1.)(x)[0, 0]))
torch.manual_seed(0)
print("RandomRotation90(1)     argmax:", argmax2d(K.RandomRotation90(times=(1, 1), p=1.)(x)[0, 0]))
torch.manual_seed(0)
aff = K.RandomAffine(degrees=(90., 90.), p=1.)
print("RandomAffine(90)        argmax:", argmax2d(aff(x)[0, 0]))

# the same split one level down, at the matrix layer
c = torch.tensor([[3., 3.]]); s = torch.ones(1, 2); a = torch.tensor([90.])
print("get_rotation_matrix2d(+90) + warp_affine argmax:",
      argmax2d(warp_affine(x, get_rotation_matrix2d(c, a, s), (7, 7))[0, 0]))
print("get_affine_matrix2d(angle=+90) + warp_affine argmax:",
      argmax2d(warp_affine(x, get_affine_matrix2d(torch.zeros(1, 2), c, s, a)[:, :2], (7, 7))[0, 0]))
print("kornia.geometry.transform.rotate(+90) argmax:", argmax2d(rotate(x, a, center=c)[0, 0]))

# shear and translate are NOT negated by the same composer
torch.manual_seed(0); msh = K.RandomShear((30., 30.), p=1.); msh(x)
torch.manual_seed(0); mafs = K.RandomAffine(degrees=0., shear=(30., 30.), p=1.); mafs(x)
print("RandomShear(30) M       :", msh.transform_matrix[0, :2].tolist())
print("RandomAffine(shear=30) M:", mafs.transform_matrix[0, :2].tolist())
torch.manual_seed(0); maft = K.RandomAffine(degrees=0., translate=(0.25, 0.25), p=1.); maft(x)
print("RandomAffine(translate) drawn:", [round(v, 6) for v in maft._params["translations"][0].tolist()],
      " M[0][2], M[1][2]:", [round(maft.transform_matrix[0, 0, 2].item(), 6),
                             round(maft.transform_matrix[0, 1, 2].item(), 6)])

# 3D: the same mirror, on every axis
V = torch.zeros(1, 1, 7, 9, 11); V[0, 0, 3 + 1, 4 + 2, 5 + 3] = 1.0  # centre (3,4,5) + (1,2,3)
def off3(t):
    i = int(t.flatten().argmax()); d, h, w = t.shape[-3:]
    z = i // (h * w); y = (i % (h * w)) // w; xx = i % w
    return (z - 3, y - 4, xx - 5)
for slot, name in ((0, "yaw  (degrees[0])"), (1, "pitch(degrees[1])"), (2, "roll (degrees[2])")):
    deg = [(0., 0.)] * 3; deg[slot] = (90., 90.)
    torch.manual_seed(0); r3 = K.RandomRotation3D(degrees=tuple(deg), p=1.)(V)
    torch.manual_seed(0); a3 = K.RandomAffine3D(degrees=tuple(deg), p=1.)(V)
    ang = torch.zeros(1, 3); ang[0, slot] = 90.
    prim = rotate3d(V, ang[:, 0], ang[:, 1], ang[:, 2])
    print(f"{name}: rotate3d {off3(prim[0,0])}  RandomRotation3D {off3(r3[0,0])}  RandomAffine3D {off3(a3[0,0])}")
```

## Observed

```
kornia 0.9.0rc1 torch 2.14.0
RandomRotation(90)      argmax: (2, 1)
RandomRotation90(1)     argmax: (2, 1)
RandomAffine(90)        argmax: (4, 5)
get_rotation_matrix2d(+90) + warp_affine argmax: (2, 1)
get_affine_matrix2d(angle=+90) + warp_affine argmax: (4, 5)
kornia.geometry.transform.rotate(+90) argmax: (2, 1)
RandomShear(30) M       : [[1.0, -0.5773503184318542, 1.732050895690918], [-0.0, 1.0, 0.0]]
RandomAffine(shear=30) M: [[1.0, -0.5773503184318542, 1.732050895690918], [0.0, 1.0, 0.0]]
RandomAffine(translate) drawn: [0.938776, -1.440329]  M[0][2], M[1][2]: [0.938776, -1.440329]
yaw  (degrees[0]): rotate3d (2, -1, 3)  RandomRotation3D (2, -1, 3)  RandomAffine3D (-2, 1, 3)
pitch(degrees[1]): rotate3d (-3, 2, 1)  RandomRotation3D (-3, 2, 1)  RandomAffine3D (3, 2, -1)
roll (degrees[2]): rotate3d (1, 3, -2)  RandomRotation3D (1, 3, -2)  RandomAffine3D (1, -3, 2)
```

## Expected

Two coherent outcomes:

1. **One direction across the package.** The canonical rule this batch documents is *positive angle = counter-clockwise as displayed*, which is what the seven other entry points already do. Adopting it means dropping the negation in `get_affine_matrix2d` and `get_affine_matrix3d`, which silently changes the sign of every rotation that `RandomAffine`, `RandomAffine3D` and any direct caller of those two composers produces today.
2. **Keep the negation and document it on every class that inherits it**, so `RandomAffine` and `RandomAffine3D` state in their own `Args` that a positive `degrees` turns the opposite way to `RandomRotation` / `RandomRotation3D`.

The audit recommends (1): the composers are the minority, the flip is one sign per function, and the split is invisible to a caller who never compares the two classes side by side. It is a behavior change for existing users, so it needs a deprecation path — a `angle_convention=` argument or a release-noted flip — rather than a silent correction. The fix scope is **angle-only**: the observed output above shows the same composer passing shear and translate through unnegated, so a patch that flips those as well would introduce a second defect.

## Classification

**Semantic contract** — two contradictory rotation-sign conventions inside one package, each internally consistent, neither stated on the classes that inherit the minority one. Idiom: none of 1–7; candidate 8: a composer negates a parameter relative to the primitive it delegates to, and every class built on it inherits the flip undocumented.

## Disposition

`MERGE IN WINDOW — behavior change`. One issue for the 2D and 3D halves, because they are one root cause: `get_affine_matrix2d` negates `angle` and `get_affine_matrix3d` negates `angles`, and `RandomAffine` / `RandomAffine3D` are the classes that expose it. Batch 6 documents the current directions on both classes, points at the batch-1 `Convention:` blocks that already record the negation at the primitives, and pins the executed argmax literals; it does not change the sign. A PR that flips the direction ahead of the coordinated Repair & Deprecation window will be closed and pointed here.

## Environment

kornia 0.9.0rc1 @ `d6a2c138` (main), torch 2.14.0, python 3.11, cpu. Found during the augmentation conventions audit (batch 6, tracking issue #4407); the batch documents the behavior above as it is and pins it, and does not change it.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

**Comment by ducha-aiki** [comment]
A measurement that qualifies this issue's premise for the 3D classes (found while reviewing #4634; `main` @ ff204fb9c, torch 2.14.0, CPU). A marker at `(row 1, col 4)` of a 7×7 slice, +90° (roll for the 3D entry points, middle depth slice read back):

| entry point | marker lands at | displayed direction |
|---|---|---|
| `kornia.geometry.transform.rotate` | (2, 1) | counter-clockwise |
| `RandomRotation` | (2, 1) | counter-clockwise |
| `RandomAffine3D` | (2, 1) | counter-clockwise |
| `RandomAffine` | (4, 5) | clockwise |
| `RandomRotation3D` | (4, 5) | clockwise |
| `kornia.geometry.transform.rotate3d` | (4, 5) | clockwise |

So in 2D `RandomAffine` is the odd one out, as the title says, but in 3D it is `RandomRotation3D` and `rotate3d` that disagree with `rotate`, and `RandomAffine3D` that agrees with it. Whatever direction is chosen as the convention, the 3D half of the fix is not "flip `RandomAffine3D`". #4634 now documents the measured directions and pins them in `test_wart_positive_roll_direction_splits_the_rotation_entry_points_4408`.

Posted on behalf of @ducha-aiki by Claude (claude-fable-5-1).

**Comment by ducha-aiki** [comment]
**Maintainer ruling (2026-09-21, at the #4407 close-out): outcome (1).** A positive angle turns the image **counter-clockwise as displayed** on the affine composer family as well; the clockwise direction is repaired, not documented.

**This issue is reserved for a maintainer implementation.** It is not open for contribution: the change flips the sign of every asymmetric `RandomAffine(degrees=...)` draw for existing users, has to land inside the coordinated Repair & Deprecation window (after `0.9.0rc2`, before `0.9.0` final) together with its `breaking` changelog fragment, and moves pins in three test files. A PR against this issue from outside the maintainer team will be closed and pointed here.

Re-measured on `main` @ `c90c05f66` (torch 2.14.0, cpu). Marker at centre + (dr, dc) = (−2, +1), +90°; counter-clockwise as displayed gives (−1, −2), clockwise gives (+1, +2):

| entry point | result |
|---|---|
| `rotate`, the `Affine` module, `RandomRotation` | (−1, −2) — counter-clockwise |
| `RandomAffine`, `get_affine_matrix2d` + `warp_affine` | (+1, +2) — clockwise |

**Phase 1 — 2D (ruled).**
- `get_affine_matrix2d` stops negating `angle` before `get_rotation_matrix2d`. Angle only: shear and translate are not negated today and stay as they are.
- `Affine.forward` passes `-self.angle` into the composer today to cancel that negation (`kornia/geometry/transform/affwarp.py:978`); it becomes `self.angle`, and the module's displayed direction must not change — that cancellation gets its own pin.
- `RandomAffine` then agrees with `RandomRotation`: same seed and same `degrees` give the same image and the same `transform_matrix`. `test_wart_random_affine_rotation_sign_4408` is inverted into that convention pin; the deviation line on `RandomAffine`, the exception sentence on the `AugmentationBase2D` anchor and on the Conventions page, and `get_affine_matrix2d`'s own `Convention:` block are rewritten in the same change.
- No `angle_convention=` argument. `0.9.0` is the single-migration release and a switch would outlive it. The fragment states what used to happen and the one-line migration (negate the angle, or turn `degrees=(a, b)` into `(-b, -a)`). A symmetric `degrees=d` draws from the same distribution before and after; only asymmetric ranges and direct callers of `get_affine_matrix2d` see a difference.

**Phase 2 — 3D (not ruled yet).** The earlier comment stands and extends to all three axes. With each rotation plane displayed earlier-dim-as-rows:

| slot | plane | `rotate3d` / `RandomRotation3D` | `RandomAffine3D` |
|---|---|---|---|
| yaw (`degrees[0]`) | D–H | (+1, +2) | (−1, −2) |
| pitch (`degrees[1]`) | D–W | (−1, −2) | (+1, +2) |
| roll (`degrees[2]`) | H–W | (+1, +2) — clockwise | (−1, −2) — counter-clockwise |

Each family has one handedness over the three axes (the pitch row flips only because D–W is the anti-cyclic plane) and the two families are mirrors. In 3D it is `RandomAffine3D` whose roll already matches 2D `rotate`, so "make the affine class counter-clockwise" is already true there, and the open question is whether the `get_projective_transform` / `rotate3d` / `RandomRotation3D` / `RandomMotionBlur3D` family moves to match 2D (larger blast radius, touches `kornia.geometry`), or 3D is only made internally consistent. That needs its own ruling before any 3D code changes; this issue stays open until it is made.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

**Comment by ducha-aiki** [comment]
Handedness check for the 3D half, on `main` @ `c90c05f66` (torch 2.14.0, cpu), in index coordinates (x, y, z) = (W, H, D). It corrects the framing of my previous comment.

```python
import torch
from kornia.geometry.transform import get_projective_transform, get_affine_matrix3d
c, s = torch.zeros(1, 3), torch.ones(1, 3)
a = torch.tensor([[0., 0., 90.]])                       # slot 2
print(get_projective_transform(c, a, s)[0, :, :3].round())   # [[0,-1,0],[1,0,0],[0,0,1]]  = Rz(+90)
print(get_affine_matrix3d(torch.zeros(1, 3), c, s, a)[0, :3, :3].round())  # its transpose
```

- `get_projective_transform` returns the textbook right-handed matrices, det +1: slot 0 is Rx, slot 1 is Ry, slot 2 is Rz. (So `yaw` / `pitch` / `roll` name rotations about x / y / z.)
- `rotate3d` and `RandomRotation3D` move image content by exactly that matrix. A marker on +y goes to +z for slot 0, +z goes to +x for slot 1, +x goes to +y for slot 2: **right-handed on all three axes**.
- `get_affine_matrix3d` and `RandomAffine3D` give the transposes: **left-handed on all three axes**.
- 2D `rotate(+90)` sends +x to −y. With y pointing down that is left-handed about +z; it is the screen convention, counter-clockwise as displayed.

This is the contract the Conventions page already states: 3D rotations follow the right-hand rule, a positive rotation about +z therefore turns clockwise on screen, the opposite of `rotate(img, +angle)`, and the two are not to be mixed without negating. "Roll looks like 2D" and "right-handed in (x, y, z)" cannot both hold while y points down.

So `RandomAffine3D` matching 2D `rotate` on screen is not a point in its favour: it is the one 3D entry point that breaks the documented right-hand rule. The 3D fix that agrees with the rest of the library is the mirror of phase 1 — `get_affine_matrix3d` stops negating `angles`, `RandomAffine3D` becomes right-handed like `rotate3d` and `RandomRotation3D`, and `get_projective_transform`, `rotate3d`, `RandomRotation3D` and `RandomMotionBlur3D` do not move. Moving the `rotate3d` family instead would make every 3D rotation in kornia left-handed and contradict `axis_angle_to_rotation_matrix`. The 3D ruling is pending the maintainer's confirmation of this reading.

Two docstring defects for the same change: `rotate3d`'s summary line says "anti-clockwise", which is not what a positive roll does on screen, and the slot-to-axis mapping (yaw → x, pitch → y, roll → z) is stated nowhere.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

**Comment by ducha-aiki** [comment]
**Maintainer ruling for the 3D half (2026-09-21): kornia's 3D rotations stay right-handed.** `get_affine_matrix3d` stops negating `angles` before `get_projective_transform`, so `RandomAffine3D` becomes right-handed in (x, y, z) = (W, H, D) like `rotate3d` and `RandomRotation3D`. `get_projective_transform`, `rotate3d`, `RandomRotation3D` and `RandomMotionBlur3D` do not change. Angle only, as in 2D: shear, translate and scale are untouched.

Full scope of this issue, both halves maintainer-implemented inside the Repair & Deprecation window:

| | change | result |
|---|---|---|
| 2D | `get_affine_matrix2d` drops the negation; `Affine.forward` drops its compensating `-self.angle` | `RandomAffine` turns counter-clockwise on screen, equal to `RandomRotation` for the same draw; `Affine` unchanged |
| 3D | `get_affine_matrix3d` drops the negation | `RandomAffine3D` is right-handed, equal to `RandomRotation3D` for the same draw |

The 2D and 3D screen directions stay opposite on purpose, as the Conventions page already states: 2D follows the screen convention, 3D follows the right-hand rule, and y points down. The three `*_4408` pins in `tests/augmentation/test_conventions_3d.py` turn into convention pins on that statement, `rotate3d`'s "anti-clockwise" summary line is corrected, and the yaw / pitch / roll → x / y / z mapping gets written down. One `breaking` changelog fragment names both composers and gives the migration (negate the angle, or turn `degrees=(a, b)` into `(-b, -a)`).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
